src/mvba/vcbc/tests.rs

```diff
@@ -1,174 +1,379 @@
+            nodes,
+            queue: Default::default(),
         }
     }
 
-    pub fn propose_msg(&self) -> Vec<u8> {
-        bincode::serialize(&Message::Propose(self.proposal.clone())).unwrap()
+    fn node_mut(&mut self, id: NodeId) -> &mut Vcbc {
+        self.nodes.get_mut(&id).unwrap()
     }
 
-    pub fn echo_msg(&self) -> Vec<u8> {
-        bincode::serialize(&Message::Echo(self.proposal.clone())).unwrap()
+    fn enqueue_bundles_from(&mut self, id: NodeId) {
+        let (send_bundles, bcast_bundles) = {
+            let mut broadcaster = self.node_mut(id).broadcaster.borrow_mut();
+            let send_bundles = broadcaster.take_send_bundles();
+            let bcast_bundles = broadcaster.take_broadcast_bundles();
+            (send_bundles, bcast_bundles)
+        };
+
+        for (recipient, bundle) in send_bundles {
+            let bundle: Bundle =
+                bincode::deserialize(&bundle).expect("Failed to deserialize bundle");
+            self.queue.entry(recipient).or_default().push(bundle);
+        }
+
+        for bundle in bcast_bundles {
+            for recipient in self.nodes.keys() {
+                let bundle: Bundle =
+                    bincode::deserialize(&bundle).expect("Failed to deserialize bundle");
+                self.queue.entry(*recipient).or_default().push(bundle);
+            }
+        }
     }
 
-    pub fn is_proposed(&self) -> bool {
-        self.broadcaster.borrow().has_message(&self.propose_msg())
+    fn drain_queue(&mut self) {
+        while !self.queue.is_empty() {
+            for (recipient, queue) in std::mem::take(&mut self.queue) {
+                let recipient_node = self.node_mut(recipient);
+
+                for bundle in queue {
+                    let msg: Message = bincode::deserialize(&bundle.message)
+                        .expect("Failed to deserialize message");
+
+                    recipient_node
+                        .receive_message(bundle.sender, msg)
+                        .expect("Failed to receive msg");
+                }
+
+                self.enqueue_bundles_from(recipient);
+            }
+        }
     }
-    pub fn is_echoed(&self) -> bool {
-        self.broadcaster.borrow().has_message(&self.echo_msg())
+
+    fn deliver(&mut self, recipient: NodeId, index: usize) {
+        if let Some(msgs) = self.queue.get_mut(&recipient) {
+            if msgs.is_empty() {
+                return;
+            }
+            let index = index % msgs.len();
+
+            let bundle = msgs.swap_remove(index);
+            let msg: Message =
+                bincode::deserialize(&bundle.message).expect("Failed to deserialize message");
+
+            let recipient_node = self.node_mut(recipient);
+            recipient_node.receive_message(bundle.sender, msg);
+            self.enqueue_bundles_from(recipient);
+        }
     }
 }
 
 #[test]
-fn test_should_propose() {
-    let mut t = TestData::new(TestData::PARTY_X);
+fn test_vcbc_happy_path() {
```

Comment:
This looks like a system test for me, not a unit test. In unit tests we should test as much as possible a small unit of the code.

Reply:
Discussed in syncup: will keep these system tests here in this file until the full implementation is ready to replace the sn_consensus crate